presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -433,12 +434,13 @@ public PlanOptimizers(
                         .build()));
 
         if (!forceSingleNode) {
+            builder.add(new ReplicateSemiJoinInDelete()); // Must run before AddExchanges and before DetermineSemiJoinDistributionType
             builder.add((new IterativeOptimizer(
                     ruleStats,
                     statsCalculator,
                     estimatedExchangesCostCalculator,
-                    ImmutableSet.of(new DetermineJoinDistributionType(costComparator, taskCountEstimator))))); // Must run before AddExchanges
-            builder.add(new DetermineSemiJoinDistributionType()); // Must run before AddExchanges
+                    ImmutableSet.of(new DetermineJoinDistributionType(costComparator, taskCountEstimator), // Must run before AddExchanges
+                            new DetermineSemiJoinDistributionType(costComparator, taskCountEstimator))))); // Must run before AddExchanges and after ReplicateSemiJoinInDelete
```

Comment:
> and after ReplicateSemiJoinInDelete

add explanation why is that. Something like "otherwise it would temporarily produce an implementable plan" or something more in English